File: ios/BidmadPluginCommonModule.m
_Pragma("clang diagnostic push")
_Pragma("clang diagnostic ignored \"-Wstrict-prototypes\"")

#import <React/RCTBridgeModule.h>
#import <OpenBiddingHelper/OpenBiddingHelper.h>

@interface RCT_EXTERN_MODULE(BidmadPluginCommonModule, NSObject)

_RCT_EXTERN_REMAP_METHOD(initializeSdk, initializeSdkWithAppKey:(NSString *)appKey resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(reqAdTrackingAuthorization, reqAdTrackingAuthorizationWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(setAdvertiserTracking, setAdvertiserTrackingWithEnable:(BOOL)enable, false)
_RCT_EXTERN_REMAP_METHOD(advertiserTracking, getAdvertiserTrackingWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(setIsChildDirectedAds, setWithIsChildDirectedAds:(BOOL)isChildDirectedAds, false)
_RCT_EXTERN_REMAP_METHOD(isChildDirectedTreatment, isChildDirectedTreatmentWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(setUserConsentCCPA, setWithUserConsentStatusForCCPACompliance:(BOOL)userConsentStatusForCCPACompliance, false)
_RCT_EXTERN_REMAP_METHOD(isUserConsentCCPA, isUserConsentCCPAWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(setIsDebug, setWithIsDebug:(BOOL)isDebug, false)
_RCT_EXTERN_REMAP_METHOD(isDebug, isDebugWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(setTestDeviceId, setWithTestDeviceId:(NSString *)testDeviceId, false)
_RCT_EXTERN_REMAP_METHOD(testDeviceId, testDeviceIdWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(setCuid, setWithCuid:(NSString *)cuid, false)
_RCT_EXTERN_REMAP_METHOD(cuid, cuidWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)
_RCT_EXTERN_REMAP_METHOD(setUseServerSideCallback, setWithUseServerSideCallback:(BOOL)useServerSideCallback, false)
_RCT_EXTERN_REMAP_METHOD(useServerSideCallback, useServerSideCallbackWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject, false)

@end

_Pragma("clang diagnostic pop")

File: ios/BidmadPluginCommonModule.swift
//
//  BidmadCommon.swift
//  react-native-bidmad-plugin-test
//
//  Created by 플랫폼-오승섭-맥북 on 2023/11/30.

//

import Foundation
import OpenBiddingHelper

@objcMembers
@objc(BidmadPluginCommonModule)
class BidmadPluginCommonModule: NSObject, RCTBridgeModule {
    
    static func requiresMainQueueSetup() -> Bool {
        return false
    }
    
    static func moduleName() -> String! {
        return "BidmadPluginCommonModule"
    }
    
    func initializeSdk(appKey: String,
                       resolver resolve: @escaping RCTPromiseResolveBlock,
                       rejecter reject: @escaping RCTPromiseRejectBlock) {
        BidmadCommon.initializeSdk(withAppKey: appKey) { success in
            if success {
                resolve(true)
            } else {
                reject("E_INIT_FAILED", "Initialization failed", nil)
            }
        }
    }
    
    func reqAdTrackingAuthorization(resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        BidmadCommon.reqAdTrackingAuthorization { status in
            resolve(status.rawValue)
        }
    }
    
    func setAdvertiserTracking(enable: Bool) {
        BidmadCommon.setAdvertiserTrackingEnabled(enable)
    }
    
    func getAdvertiserTracking(resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        resolve(BidmadCommon.getAdvertiserTrackingEnabled())
    }
    
    func set(isChildDirectedAds: Bool) {
        BidmadCommon.setIsChildDirectedAds(isChildDirectedAds)
    }
    
    func isChildDirectedTreatment(resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        guard let isChildDirectedAds = BidmadCommon.isChildDirectedTreatment() else {
            resolve(nil)
            return
        }
        
        resolve(isChildDirectedAds.boolValue)
    }
    
    func set(userConsentStatusForCCPACompliance: Bool) {
        BidmadCommon.setUserConsentStatusForCCPACompliance(userConsentStatusForCCPACompliance)
    }
    
    func isUserConsentCCPA(resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        guard let isUserConsentCCPA = BidmadCommon.isUserConsentCCPA() else {
            resolve(nil)
            return
        }
        
        resolve(isUserConsentCCPA.boolValue)
    }
    
    func set(isDebug: Bool) {
        BidmadCommon.setIsDebug(isDebug)
    }
    
    func isDebug(resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        resolve(BidmadCommon.isDebug())
    }
    
    func set(testDeviceId: String) {
        BidmadCommon.setTestDeviceId(testDeviceId)
    }
    
    func testDeviceId(resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        resolve(BidmadCommon.testDeviceId())
    }

    func set(cuid: String) {
        BidmadCommon.setCuid(cuid)
    }
    
    func cuid(resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        resolve(BidmadCommon.cuid())
    }
    
    func set(useServerSideCallback: Bool) {
        BidmadCommon.setUseServerSideCallback(useServerSideCallback)
    }
    
    func useServerSideCallback(resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        resolve(BidmadCommon.useServerSideCallback())
    }
}
